lib/hcrypto/evp.c

```diff
@@ -857,60 +871,72 @@ EVP_CipherUpdate(EVP_CIPHER_CTX *ctx, void *out, int *outlen,
 
     *outlen = 0;
 
-    /**
-     * If there in no spare bytes in the left from last Update and the
-     * input length is on the block boundery, the EVP_CipherUpdate()
-     * function can take a shortcut (and preformance gain) and
-     * directly encrypt the data, otherwise we hav to fix it up and
-     * store extra it the EVP_CIPHER_CTX.
+    /*
+     * If there in no bytes left over from the last Update and the
+     * input length is on a block boundary, then we can take a
+     * shortcut (and preformance gain) and directly encrypt the
+     * data.
      */
-    if (ctx->buf_len == 0 && (inlen & ctx->block_mask) == 0) {
+    if (ctx->buf_len == 0 && inlen && (inlen & ctx->block_mask) == 0) {
+        /*
+         * OpenSSL's EVP_Cipher() returns the number of bytes on success, 0 on
+         * failure.
+         */
 	ret = (*ctx->cipher->do_cipher)(ctx, out, in, inlen);
-	if (ret == 1)
+	if (ret == 1 || ret == inlen) {
 	    *outlen = inlen;
-	else
+            ret = 1;
+        } else
 	    *outlen = 0;
 	return ret;
     }
 
-
     blocksize = EVP_CIPHER_CTX_block_size(ctx);
     left = blocksize - ctx->buf_len;
     assert(left > 0);
 
     if (ctx->buf_len) {
-
-	/* if total buffer is smaller then input, store locally */
-	if (inlen < left) {
-	    memcpy(ctx->buf + ctx->buf_len, in, inlen);
-	    ctx->buf_len += inlen;
-	    return 1;
-	}
-
-	/* fill in local buffer and encrypt */
-	memcpy(ctx->buf + ctx->buf_len, in, left);
-	ret = (*ctx->cipher->do_cipher)(ctx, out, ctx->buf, blocksize);
-	memset_s(ctx->buf, blocksize, 0, blocksize);
-	if (ret != 1)
-	    return ret;
-
-	*outlen += blocksize;
-	inlen -= left;
-	in = ((unsigned char *)in) + left;
-	out = ((unsigned char *)out) + blocksize;
-	ctx->buf_len = 0;
+        /* If we can't fill one block in the buffer, save the input there */
+        if (inlen < left) {
+            memcpy(ctx->buf + ctx->buf_len, in, inlen);
+            ctx->buf_len += inlen;
+            return 1;
+        }
+
+        /* Fill the buffer and encrypt */
+        memcpy(ctx->buf + ctx->buf_len, in, left);
+        ret = (*ctx->cipher->do_cipher)(ctx, out, ctx->buf, blocksize);
+        memset_s(ctx->buf, blocksize, 0, blocksize);
+        /*
+         * OpenSSL's EVP_Cipher() returns the number of bytes on success, 0 on
+         * failure.
```

Comment:
do_cipher() cannot be EVP_Cipher().   ctx->cipher->do_cipher() is called from EVP_Cipher().